### PR TITLE
Force consumers of Connection::waitForAndDispatchImmediately to declare usage in the messages.in file

### PR DESCRIPTION
**Pull request by mattwoodrow**
#### 2c3c1f0afc15e23fcac517543fe80b8c43bcdfaf
<pre>
Force consumers of Connection::waitForAndDispatchImmediately to declare usage in the messages.in file
<a href="https://bugs.webkit.org/show_bug.cgi?id=260053">https://bugs.webkit.org/show_bug.cgi?id=260053</a>

Reviewed by Kimmo Kinnunen.

waitForAndDispatchImmediately causes the waited on message to be delivered synchronous, and breaks expectations about IPC messages being delivered in order.

Usage of waitForAndDispatchImmediately can be subtle (if not explicitly searched for), so it&apos;s easy to add new IPC messages that expect in-order delivery.

This adds required &apos;CanDispatchOutOfOrder&apos; and &apos;ReplyCanDispatchOutOfOrder&apos; annotations to the messages.in files, before messages can be used with
waitForAndDispatchImmediately.

* Source/WebKit/Platform/IPC/Connection.h:
(IPC::Connection::waitForAndDispatchImmediately):
(IPC::Connection::waitForAsyncReplyAndDispatchImmediately):
* Source/WebKit/Scripts/webkit/messages.py:
(message_to_struct_declaration):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.messages.in:
* Source/WebKit/UIProcess/WebPageProxy.messages.in:
* Source/WebKit/WebProcess/GPU/GPUProcessConnection.messages.in:
* Source/WebKit/WebProcess/GPU/graphics/RemoteGraphicsContextGLProxy.messages.in:
* Source/WebKit/WebProcess/GPU/graphics/RemoteImageBufferProxy.messages.in:
* Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.messages.in:
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteGPUProxy.messages.in:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
* Source/WebKit/WebProcess/WebProcess.messages.in:
* Tools/TestWebKitAPI/Tests/IPC/IPCTestUtilities.h:

Canonical link: <a href="https://commits.webkit.org/267178@main">https://commits.webkit.org/267178@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/df1759bded9c3af7ff92eb00bad3d0bd54f9a0ba

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/15865 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/16184 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/16566 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/17628 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/14891 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/16051 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/19194 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/16282 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/17358 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/16056 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/16527 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/13515 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/18386 "Built successfully") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/15983 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/13776 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/14332 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/21214 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/14768 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/14497 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/17747 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/15093 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/12793 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/14334 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/3789 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/18703 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/14915 "Built successfully") | | |

### CODE REVIEW

**Review by kkinnunen-apple — COMMENTED** [review]
I suppose the general idea looks good.

> Usage of waitForAndDispatchImmediately can be subtle (if not explicitly searched for), so it's easy to add new IPC messages that expect in-order delivery.

I have hard time deducing the actual problem that is being solved here. Maybe rephrase?
My thinking is that when adding IPC messages that expect in-order delivery, what does waitForAndDispatchImmediately have to do with it? Presumably there is no waitForAndDispatchImmediately on these new messages?

The problem with nomenclature:
1. CanDispatchOutOfOrder, ReplyCanDispatchOutOfOrder in a message insinuates: this message may be delivered out of order in general.

The reordering happens due to what the caller does, not how the system delivers the message..

2. The naming clashes with the other features of DispatchMessageEvenWhenWaitingForSyncReply, MaintainOrderingWithAsyncMessages that are used to ensure in-order delivery.
So if user knows about this reordering, then the naming kind of conflicts with the other feature: "I thought with my flag the ordering is guaranteed, but now this other
attribute says the ordering is not guaranteed".

Initially I'd think one would call this feature "Waitable" and "ReplyWaitable". But it's not entirely clear to me what problem you are solving. 
Do you want to make the future reader know there's funny business about the message (Waitable)?
or
Do you want the initial author to reconsider their choices (CanDispatchOutOfOrder ...WarningYourCodeIsProbablyBuggyAndDoesNotHandleTheCasesNeededEvenIfYouTried.)?


I wish we would look and find some sort of a better solution to the waits..

**Review by kkinnunen-apple — APPROVED** [review]

**Comment by mattwoodrow** [comment]
> I have hard time deducing the actual problem that is being solved here. Maybe rephrase? My thinking is that when adding IPC messages that expect in-order delivery, what does waitForAndDispatchImmediately have to do with it? Presumably there is no waitForAndDispatchImmediately on these new messages?

The bug that made me want to add this had an existing IPC message (`RemoteLayerTreeDrawingAreaProxy::CommitLayerTree`), with a (relatively rare) call-site that used `waitForAndDispatchImmediately`. I added a new IPC message, where the receiver implementation depended on the relative ordering between my new message and the existing one being preserved. This caused a very infrequent race that was hard to debug.

The assertion that I am making is that this would have been easier to avoid (at code authoring, or review time), had it been more obvious that relative ordering of the messages wasn't guaranteed.

> 
> The problem with nomenclature:
> 
> 1. CanDispatchOutOfOrder, ReplyCanDispatchOutOfOrder in a message insinuates: this message may be delivered out of order in general.
> 
> The reordering happens due to what the caller does, not how the system delivers the message..
> 
> 2. The naming clashes with the other features of DispatchMessageEvenWhenWaitingForSyncReply, MaintainOrderingWithAsyncMessages that are used to ensure in-order delivery.
>    So if user knows about this reordering, then the naming kind of conflicts with the other feature: "I thought with my flag the ordering is guaranteed, but now this other
>    attribute says the ordering is not guaranteed".
> 
> Initially I'd think one would call this feature "Waitable" and "ReplyWaitable". 

I agree that these names more accurately reflect what the actual capability is, but I was trying to focus the name on what the relevant implications of that were.

As far as I'm aware, there's no documentation of what 'waitable' would mean, other than looking in Connection.h for a string that seems close. I guess the name `waitForAndDispatchImmediately` is trying to self-document, but it's still not obvious to me that it would only immediately dispatch the waited-on message (and thus re-order) and not also dispatch all the intermediate queued messages. Having to dig into the implementation to understand that isn't ideal.

Arguably that's a totally separate problem, and we should be writing decent documentation on the guarantees provided by `messages.in` and all the exceptions.

>But it's not entirely clear to me what problem you are solving. Do you want to make the future reader know there's funny business about the message (Waitable)? or Do you want the initial author to reconsider their choices (CanDispatchOutOfOrder ...WarningYourCodeIsProbablyBuggyAndDoesNotHandleTheCasesNeededEvenIfYouTried.)?

It's really both :) It's a 'here be dragons' label, both to alert engineers to existing danger, and to discourage adding more.

> 
> I wish we would look and find some sort of a better solution to the waits..

I could not agree more :)

**Comment by kkinnunen-apple** [comment]
> The assertion that I am making is that this would have been easier to avoid (at code authoring, or review time), had it been more obvious that relative ordering of the messages wasn't guaranteed.

Ok, yeah, I can see that you'd want it to be a warning not a statement of capability
